File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    Rails.logger.info "connected"
    stream_from "chat_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    Rails.logger.info "speak: #{data}"
    ActionCable.server.broadcast("chat_channel", {message: data["message"]})
  end
end

File: app/javascript/channels/chat_channel.js
import consumer from "channels/consumer"

consumer.subscriptions.create("ChatChannel", {
  connected() {
    console.log("Connected to the chat channel")
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    const messages = document.getElementById('messages')
    messages.innerHTML += `<p>${data.message}</p>`
  },

  speak(message) {
    console.log('speak', message)
    this.perform('speak', { message: message })
  }
});

document.addEventListener('DOMContentLoaded', () => {
  const input = document.getElementById('chat-input')
  input.addEventListener('keypress', (event) => {
    if (event.key === 'Enter') {
      const message = input.value
      input.value = ''
      console.log("consumer", consumer)
      consumer.subscriptions.subscriptions[0].speak(message)
    }
  })
});
